Add a sample for adding git commit id to log entries

Deren-Liao opened the pull request:
This sample will be linked by the user documentation for Logs Viewer feature inside our Google Cloud Tools for Visual Studio extension.
The main purpose is to help user to know how to add the git revision id if they decide to use lower level logging API Google.Cloud.Logging.V2 instead of using Log4Net.

SurferJeffAtGoogle commented on logging/add-git-id/App_Start/RouteConfig.cs:
```diff
@@ -0,0 +1,39 @@
+﻿/*
```
Move the code into
/logging/api/WithGitId/...

SurferJeffAtGoogle commented on logging/add-git-id/Controllers/LoggingSampleController.cs:
```diff
@@ -0,0 +1,51 @@
+            {
+                for (int i = 0; i < count; ++i)
+                {
+                    StackdriverLogWritter.WriteLog("Test. Replace with your own message.");
```
Is there a type in the name LogWritter?  Should it be LogWriter?

Make the text something like "Hello Stackdriver."

SurferJeffAtGoogle commented on logging/add-git-id/Controllers/LoggingSampleController.cs:
```diff
@@ -0,0 +1,51 @@
+            }
+            else
+            {
+                throw new ArgumentOutOfRangeException("Pleae limit count in range of 1 ~ 100");
```
Or maybe default to a small number like 3?

Deren-Liao replied:
This is protective code. The UI already set limit to 100.   Just in case somebody hack the input (which not likely), this protect the site creating huge number of log entries.

SurferJeffAtGoogle commented on logging/add-git-id/Global.asax.cs:
```diff
@@ -0,0 +1,18 @@
+﻿using System;
```
Add copyright.

SurferJeffAtGoogle commented on logging/add-git-id/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,35 @@
+﻿using System.Reflection;
```
Add copyright.

SurferJeffAtGoogle commented on logging/add-git-id/StackdriverLogWritter.cs:
```diff
@@ -0,0 +1,102 @@
+        private const string LogId = "YOUR-LOG-ID";
+        private const string ProjectId = "YOUR-PROJECT-ID";
+
+        [MethodImpl(methodImplOptions: MethodImplOptions.NoInlining)]
```
Why?

Deren-Liao replied:
I am NOT sure if this is inline , then would the caller information be messed up

SurferJeffAtGoogle commented on logging/add-git-id/StackdriverLogWritter.cs:
```diff
@@ -0,0 +1,102 @@
+                || ex is InvalidJsonException
+                || ex is UnauthorizedAccessException)
+            {
+                // This is best-effort only, exceptions from reading/parsing the source_context.json are ignored.
```
If this fails, then this sample fails to show users what is was designed to show them, no?  In that case, exceptions should not be ignored.  Probably the simplest option is to not catch them at all.

Deren-Liao replied:
If user incorporate the sample into their production code, I don't want the logger blow up their products.   I added Debug.Writeline(..)  , hopefully it helps in some extent.

SurferJeffAtGoogle replied:
'tis noble to not blow up user products.  :-)

I tried downloading and running the sample.  I set and the project id and log name and pressed F5.  But my logging statements contain no git revision ids.  I found it hard to find the output of Debug.Writeline().  I think many users will not see it at all.

If you still think an exception is too dangerous, can we use a Debug.Assert() instead?  It won't execute at all in a production build, but will halt the program in a debug build.

Deren-Liao commented:
PTAL.
I added a test, and README.md.  To let the test run, I upgraded the nuget packages versions so that all projects using same versions.

SurferJeffAtGoogle commented:
I got this error when I tried to build:
```
Restoring NuGet packages...
To prevent NuGet from restoring packages during build, open the Visual Studio Options dialog, click on the Package Manager node and uncheck 'Allow NuGet to download missing packages during build.'
NuGet package restore finished.
1>------ Build started: Project: WithGitId, Configuration: Debug Any CPU ------
1>Build started 5/30/2017 11:12:05 AM.
1>C:\Users\Jeffrey Rennie\gitrepos\dotnet-docs-samples\logging\api\WithGitId\WithGitId.csproj(5,3): error MSB4019: The imported project "C:\Users\Jeffrey Rennie\gitrepos\dotnet-docs-samples\logging\api\WithGitId\packages\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.1.0.0\build\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.props" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.
1>
1>Build FAILED.
1>
1>Time Elapsed 00:00:00.00
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

I searched for the error text and found:
https://stackoverflow.com/questions/32254439/nuget-packages-are-missing

SurferJeffAtGoogle commented:
Did you build this project with VS2017 or 2015?

Deren-Liao commented:
I built it with VS2015.
Just now, I tried to build with VS2015 and VS2017. I am getting this error
 --- from a fresh clone.   I mean it worked when I have the packages
locally.
Let me take a closer look.

On Tue, May 30, 2017 at 11:17 AM, Jeffrey Rennie <notifications@github.com>
wrote:

> Did you build this project with VS2017 or 2015?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/267#issuecomment-304963683>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVxZpAqdelRtCsE_XWf9IIs5UMJhG0X7ks5r_F1AgaJpZM4NkEsH>
> .
>

Deren-Liao commented:
Yes, that's the problem.  I changed the solution file path, copied project
folder over.  The packages binaries under project folder were moved
together . So it built successfully locally on my machine.

I fixed the problem and tested with clean clone on both VS2015 and VS2017

On Tue, May 30, 2017 at 11:40 AM, Deren Liao <derenl@google.com> wrote:

> I built it with VS2015.
> Just now, I tried to build with VS2015 and VS2017. I am getting this error
>  --- from a fresh clone.   I mean it worked when I have the packages
> locally.
> Let me take a closer look.
>
> On Tue, May 30, 2017 at 11:17 AM, Jeffrey Rennie <notifications@github.com
> > wrote:
>
>> Did you build this project with VS2017 or 2015?
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/267#issuecomment-304963683>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AVxZpAqdelRtCsE_XWf9IIs5UMJhG0X7ks5r_F1AgaJpZM4NkEsH>
>> .
>>
>
>

Deren-Liao commented:
I will add that.
To be clear, git id won't be added in F5.  You have to publish it so that
the source_context.json is generated.

On Tue, May 30, 2017 at 1:08 PM, Jeffrey Rennie <notifications@github.com>
wrote:

> *@SurferJeffAtGoogle* commented on this pull request.
> ------------------------------
>
> In logging/add-git-id/StackdriverLogWritter.cs
> <https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/267#discussion_r119201524>
> :
>
> > +        {
> +            try
> +            {
> +                var gitId = SourceContext.AppSourceContext?.Git?.RevisionId;
> +                if (!String.IsNullOrWhiteSpace(gitId))
> +                {
> +                    labels.Add(SourceContext.GitRevisionIdLogLabel, gitId);
> +                }
> +            }
> +            catch (Exception ex) when (
> +                ex is SecurityException
> +                || ex is InvalidProtocolBufferException
> +                || ex is InvalidJsonException
> +                || ex is UnauthorizedAccessException)
> +            {
> +                // This is best-effort only, exceptions from reading/parsing the source_context.json are ignored.
>
> 'tis noble to not blow up user products. :-)
>
> I tried downloading and running the sample. I set and the project id and
> log name and pressed F5. But my logging statements contain no git revision
> ids. I found it hard to find the output of Debug.Writeline(). I think many
> users will not see it at all.
>
> If you still think an exception is too dangerous, can we use a
> Debug.Assert() instead? It won't execute at all in a production build, but
> will halt the program in a debug build.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/267#discussion_r119201524>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVxZpJ_JfN0kli8LgGibrYaZSN_wLfVeks5r_HcxgaJpZM4NkEsH>
> .
>

Deren-Liao commented:
I added Assert.
Some users may just copy/paste the code. I really don't want a log
exception breaks their product by any means.
If you want to try get the git id added, please use our extension to
publish it to a GCE ASP.NET image VM.

Thank you for all the feedback.  I appreciate!

On Tue, May 30, 2017 at 1:10 PM, Deren Liao <derenl@google.com> wrote:

> I will add that.
> To be clear, git id won't be added in F5.  You have to publish it so that
> the source_context.json is generated.
>
> On Tue, May 30, 2017 at 1:08 PM, Jeffrey Rennie <notifications@github.com>
> wrote:
>
>> *@SurferJeffAtGoogle* commented on this pull request.
>> ------------------------------
>>
>> In logging/add-git-id/StackdriverLogWritter.cs
>> <https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/267#discussion_r119201524>
>> :
>>
>> > +        {
>> +            try
>> +            {
>> +                var gitId = SourceContext.AppSourceContext?.Git?.RevisionId;
>> +                if (!String.IsNullOrWhiteSpace(gitId))
>> +                {
>> +                    labels.Add(SourceContext.GitRevisionIdLogLabel, gitId);
>> +                }
>> +            }
>> +            catch (Exception ex) when (
>> +                ex is SecurityException
>> +                || ex is InvalidProtocolBufferException
>> +                || ex is InvalidJsonException
>> +                || ex is UnauthorizedAccessException)
>> +            {
>> +                // This is best-effort only, exceptions from reading/parsing the source_context.json are ignored.
>>
>> 'tis noble to not blow up user products. :-)
>>
>> I tried downloading and running the sample. I set and the project id and
>> log name and pressed F5. But my logging statements contain no git revision
>> ids. I found it hard to find the output of Debug.Writeline(). I think many
>> users will not see it at all.
>>
>> If you still think an exception is too dangerous, can we use a
>> Debug.Assert() instead? It won't execute at all in a production build, but
>> will halt the program in a debug build.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/267#discussion_r119201524>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AVxZpJ_JfN0kli8LgGibrYaZSN_wLfVeks5r_HcxgaJpZM4NkEsH>
>> .
>>
>
>

SurferJeffAtGoogle commented on logging/api/WithGitId/StackdriverLogWriter.cs:
```diff
@@ -0,0 +1,105 @@
+
+        private static void TryAddGitRevisionId(IDictionary<string, string> labels)
+        {
+            try
```
This code can be a little tighter:
```cs
            try
            {
                var gitId = SourceContext.AppSourceContext?.Git?.RevisionId;
                Debug.Assert(!String.IsNullOrWhiteSpace(gitId), "No valid git revision id found.",
                    "Make sure you have source-context.json published with the application.");
                labels.Add(SourceContext.GitRevisionIdLogLabel, gitId);
            }
            catch (Exception ex) when (
                ex is SecurityException
                || ex is InvalidProtocolBufferException
                || ex is InvalidJsonException
                || ex is UnauthorizedAccessException)
            {
                // This is best-effort only, exceptions from reading/parsing the source_context.json are ignored.
                Debug.Fail("Exception at TryAddGitRevisionId.", ex.ToString());
            }
```

Deren-Liao commented:
Done.
Please take a look again.

On Tue, May 30, 2017 at 4:23 PM, Jeffrey Rennie <notifications@github.com>
wrote:

> *@SurferJeffAtGoogle* commented on this pull request.
> ------------------------------
>
> In logging/api/WithGitId/StackdriverLogWriter.cs
> <https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/267#discussion_r119239516>
> :
>
> > +            };
> +
> +            TryAddGitRevisionId(entryLabels);
> +
> +            client.WriteLogEntries(
> +                logName: LogNameOneof.From(logName),
> +                resource: resource,
> +                labels: entryLabels,
> +                entries: new[] { logEntry },
> +                callSettings: null);
> +            return logEntry.ToString();
> +        }
> +
> +        private static void TryAddGitRevisionId(IDictionary<string, string> labels)
> +        {
> +            try
>
> This code can be a little tighter:
>
>             try
>             {
>                 var gitId = SourceContext.AppSourceContext?.Git?.RevisionId;
>                 Debug.Assert(!String.IsNullOrWhiteSpace(gitId), "No valid git revision id found.",
>                     "Make sure you have source-context.json published with the application.");
>                 labels.Add(SourceContext.GitRevisionIdLogLabel, gitId);
>             }
>             catch (Exception ex) when (
>                 ex is SecurityException
>                 || ex is InvalidProtocolBufferException
>                 || ex is InvalidJsonException
>                 || ex is UnauthorizedAccessException)
>             {
>                 // This is best-effort only, exceptions from reading/parsing the source_context.json are ignored.
>                 Debug.Fail("Exception at TryAddGitRevisionId.", ex.ToString());
>             }
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/267#pullrequestreview-41100646>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVxZpN41LcwwCFQpObNBI6BVwtIC0vFkks5r_KTagaJpZM4NkEsH>
> .
>

SurferJeffAtGoogle commented:
Thank you Deren!

Deren-Liao commented:
Thanks!
